drivers/wifi/infineon/airoc_whd_hal.c

```diff
@@ -0,0 +1,474 @@
+	.whd_buffer_set_size = cy_buffer_set_size,
+	.whd_buffer_add_remove_at_front = cy_buffer_add_remove_at_front,
+};
+
+static whd_netif_funcs_t netif_if_default = {
+	.whd_network_process_ethernet_data = cy_network_process_ethernet_data,
+};
+
+/******************************************************
+ *                 Function
+ ******************************************************/
+
+int airoc_wifi_power_on(const struct device *dev)
+{
+	int result = 0;
+	const struct airoc_wifi_config *config = dev->config;
+
+#if DT_INST_NODE_HAS_PROP(0, wifi_reg_on_gpios)
+	/* Check WIFI REG_ON gpio instance */
+	if (!device_is_ready(config->wifi_reg_on_gpio.port)) {
+		LOG_ERR("Error: failed to configure wifi_reg_on %s pin %d",
+			config->wifi_reg_on_gpio.port->name, config->wifi_reg_on_gpio.pin);
+		return -EIO;
+	}
+
+	/* Configure wifi_reg_on as output  */
+	result = gpio_pin_configure_dt(&config->wifi_reg_on_gpio, GPIO_OUTPUT);
+	if (result) {
+		LOG_ERR("Error %d: failed to configure wifi_reg_on %s pin %d", result,
+			config->wifi_reg_on_gpio.port->name, config->wifi_reg_on_gpio.pin);
+		return result;
+	}
+	result = gpio_pin_set_dt(&config->wifi_reg_on_gpio, 0);
+	if (result) {
+		return result;
+	}
+
+	/* Allow CBUCK regulator to discharge */
+	(void)cyhal_system_delay_ms(WLAN_CBUCK_DISCHARGE_MS);
+
+	/* WIFI power on */
+	result = gpio_pin_set_dt(&config->wifi_reg_on_gpio, 1);
+	if (result) {
+		return result;
+	}
+	(void)cyhal_system_delay_ms(WLAN_POWER_UP_DELAY_MS);
+#endif /* DT_INST_NODE_HAS_PROP(0, reg_on_gpios) */
+
+	return result;
+}
+
+int airoc_wifi_init_primary(const struct device *dev, whd_interface_t *interface)
+{
+	int ret;
+	struct airoc_wifi_data *data = dev->data;
+	const struct airoc_wifi_config *config = dev->config;
+
+	whd_sdio_config_t whd_sdio_config = {
+		.sdio_1bit_mode = WHD_FALSE,
+		.high_speed_sdio_clock = WHD_FALSE,
+	};
+
+#if DT_INST_NODE_HAS_PROP(0, wifi_host_wake_gpios)
+	whd_oob_config_t oob_config = {
+		.host_oob_pin = (void *)&config->wifi_host_wake_gpio,
+		.dev_gpio_sel = DEFAULT_OOB_PIN,
+		.is_falling_edge =
+			(CY_WIFI_HOST_WAKE_IRQ_EVENT == GPIO_INT_TRIG_LOW) ? WHD_TRUE : WHD_FALSE,
+		.intr_priority = CY_WIFI_OOB_INTR_PRIORITY};
+	whd_sdio_config.oob_config = oob_config;
+#endif
+
+	if (airoc_wifi_power_on(dev)) {
+		LOG_ERR("airoc_wifi_power_on retuens fail");
+		return -ENODEV;
+	}
+
+	if (!device_is_ready(config->sdhc_dev)) {
+		LOG_ERR("SDHC device is not ready");
+		return -ENODEV;
+	}
+
+	ret = sd_init(config->sdhc_dev, &data->card);
+	if (ret) {
+		return ret;
+	}
+
+	cy_rslt_t result = whd_init(&data->whd_drv, &init_config_default, &resource_ops,
+				    &buffer_if_default, &netif_if_default);
+	if (result == CY_RSLT_SUCCESS) {
+		result = whd_bus_sdio_attach(data->whd_drv, &whd_sdio_config,
+					     (whd_sdio_t)&data->card);
+
+		if (result == CY_RSLT_SUCCESS) {
+			result = whd_wifi_on(data->whd_drv, interface);
+		}
+
+		if (result != CY_RSLT_SUCCESS) {
+			whd_deinit(*interface);
```

Comment:
Should we also deinit the SDIO if this fails? If so, this whole if statement after whd_init should have an else statement to deinit the SDIO in case whd_init fails.

Reply:
Zephyr does't  hve deinit function for sd in sd.c